src/graph_gui.cpp

```diff
@@ -1329,6 +1332,267 @@ struct PerformanceRatingDetailWindow : Window {
+		for (size_t j = 0; j < INDUSTRY_NUM_OUTPUTS; j++) {
+			if (!IsCargoIDValid(i->produced_cargo[j])) continue;
+			count++;
+		}
+		this->vscroll->SetCount(count);
+
+		this->FinishInitNested(window_number);
+
+		/* Initialise the dataset */
+		this->UpdateStatistics(true);
+	}
+
+	void UpdateExcludedData()
+	{
+		this->excluded_data = 0;
+
+		int index = 0;
+		const Industry *i = Industry::Get(this->window_number);
+		const CargoSpec *cs;
+		for (size_t j = 0; j < INDUSTRY_NUM_OUTPUTS; j++) {
+			if (!IsCargoIDValid(i->produced_cargo[j])) continue;
+
+			cs = CargoSpec::Get(i->produced_cargo[j]);
+			if (HasBit(_legend_excluded_cargo, cs->Index())) SetBit(this->excluded_data, index);
+			index++;
+		}
+	}
+
+	void UpdateWidgetSize(int widget, Dimension *size, const Dimension &padding, Dimension *fill, Dimension *resize) override
+	{
+		if (widget != WID_IP_MATRIX) {
+			BaseGraphWindow::UpdateWidgetSize(widget, size, padding, fill, resize);
+			return;
+		}
+
+		const Industry *i = Industry::Get(this->window_number);
+		const CargoSpec *cs;
+		for (size_t j = 0; j < INDUSTRY_NUM_OUTPUTS; j++) {
+			if (!IsCargoIDValid(i->produced_cargo[j])) continue;
+
+			cs = CargoSpec::Get(i->produced_cargo[j]);
+			SetDParam(0, cs->name);
+			Dimension d = GetStringBoundingBox(STR_GRAPH_CARGO_PAYMENT_CARGO);
+			d.width += 14; // colour field
+			d.width += WidgetDimensions::scaled.framerect.Horizontal();
+			d.height += WidgetDimensions::scaled.framerect.Vertical();
+			*size = maxdim(d, *size);
+		}
+
+		this->line_height = size->height;
+		size->height = this->line_height * 11; /* Default number of cargo types in most climates. */
+		resize->width = 0;
+		resize->height = this->line_height;
+	}
+
+	void DrawWidget(const Rect &r, int widget) const override
+	{
+		if (widget != WID_IP_MATRIX) {
+			BaseGraphWindow::DrawWidget(r, widget);
+			return;
+		}
+
+		bool rtl = _current_text_dir == TD_RTL;
+
+		int x = r.left + WidgetDimensions::scaled.framerect.left;
+		int y = r.top;
+
+		int pos = this->vscroll->GetPosition();
+		int max = pos + this->vscroll->GetCapacity();
+
+		const Industry *i = Industry::Get(this->window_number);
+		const CargoSpec *cs;
+		for (size_t j = 0; j < INDUSTRY_NUM_OUTPUTS; j++) {
+			if (!IsCargoIDValid(i->produced_cargo[j])) continue;
+
+			if (pos-- > 0) continue;
+			if (--max < 0) break;
+
+			cs = CargoSpec::Get(i->produced_cargo[j]);
+
+			bool lowered = !HasBit(_legend_excluded_cargo, cs->Index());
+
+			/* Redraw box if lowered */
+			if (lowered) DrawFrameRect(r.left, y, r.right, y + this->line_height - 1, COLOUR_ORANGE, lowered ? FR_LOWERED : FR_NONE);
+
+			byte clk_dif = lowered ? 1 : 0;
+			int rect_x = clk_dif + (rtl ? r.right - 12 : r.left + WidgetDimensions::scaled.framerect.left);
+
+			GfxFillRect(rect_x, y + clk_dif, rect_x + 8, y + 5 + clk_dif, PC_BLACK);
+			GfxFillRect(rect_x + 1, y + 1 + clk_dif, rect_x + 7, y + 4 + clk_dif, cs->legend_colour);
+			SetDParam(0, cs->name);
```

Comment:
There's a cleaner implementation of this in `PaymentRatesGraphWindow::DrawWidget` which should be used instead.